pkg/controller/backup_repository_controller.go

```diff
@@ -315,16 +316,41 @@ func (r *BackupRepoReconciler) runMaintenanceIfDue(ctx context.Context, req *vel
 	if err := r.repositoryManager.PruneRepo(req); err != nil {
 		log.WithError(err).Warn("error pruning repository")
 		return r.patchBackupRepository(ctx, req, func(rr *velerov1api.BackupRepository) {
-			rr.Status.Message = err.Error()
+			updateRepoMaintenanceHistory(rr, velerov1api.BackupRepositoryMaintenanceFailed, startTime, r.clock.Now(), err.Error())
 		})
 	}
 
 	return r.patchBackupRepository(ctx, req, func(rr *velerov1api.BackupRepository) {
-		rr.Status.Message = ""
-		rr.Status.LastMaintenanceTime = &metav1.Time{Time: now}
+		completionTime := r.clock.Now()
+		rr.Status.LastMaintenanceTime = &metav1.Time{Time: completionTime}
+		updateRepoMaintenanceHistory(rr, velerov1api.BackupRepositoryMaintenanceSucceeded, startTime, completionTime, "")
 	})
 }
 
+func updateRepoMaintenanceHistory(repo *velerov1api.BackupRepository, result velerov1api.BackupRepositoryMaintenanceResult, startTime time.Time, completionTime time.Time, message string) {
+	length := defaultMaintenanceStatusQueueLength
+	if len(repo.Status.RecentMaintenanceStatus) < defaultMaintenanceStatusQueueLength {
+		length = len(repo.Status.RecentMaintenanceStatus) + 1
+	}
+
+	lru := make([]velerov1api.BackupRepositoryMaintenanceStatus, length)
+
+	if len(repo.Status.RecentMaintenanceStatus) >= defaultMaintenanceStatusQueueLength {
+		copy(lru[:length-1], repo.Status.RecentMaintenanceStatus[len(repo.Status.RecentMaintenanceStatus)-defaultMaintenanceStatusQueueLength+1:])
+	} else {
+		copy(lru[:length-1], repo.Status.RecentMaintenanceStatus[:])
+	}
+
+	lru[length-1] = velerov1api.BackupRepositoryMaintenanceStatus{
+		Result:            result,
+		StartTimestamp:    &metav1.Time{Time: startTime},
+		CompleteTimestamp: &metav1.Time{Time: completionTime},
+		Message:           message,
+	}
+
+	repo.Status.RecentMaintenanceStatus = lru
+}
```

Comment:
```suggestion
func addRepoMaintenanceHistory(repo *velerov1api.BackupRepository, result velerov1api.BackupRepositoryMaintenanceResult, startTime time.Time, completionTime time.Time, message string) {
	length := defaultMaintenanceStatusQueueLength
	resultStatus := velerov1api.BackupRepositoryMaintenanceStatus{
		Result:            result,
		StartTimestamp:    &metav1.Time{Time: startTime},
		CompleteTimestamp: &metav1.Time{Time: completionTime},
		Message:           message,
	}
	startingPos := 0
	if len(repo.Status.RecentMaintenanceStatus) == length {
		// discard first element
		startingPos = 1
	}
	repo.Status.RecentMaintenanceStatus = append(repo.Status.RecentMaintenanceStatus[startingPos:], resultStatus)
}
```
Thoughts?

Reply:
This looks simpler, let me try and test it.

Reply:
Done